File: src/#889/solution.go
package solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

//pre = [1,2,4,5,3,6,7], post = [4,5,2,6,7,3,1]

func constructFromPrePost(pre []int, post []int) *TreeNode {
	if len(pre) == 0 {
		return nil
	}
	root := &TreeNode{Val: pre[0]}
	if len(pre) == 1 {
		return root
	}
	pre = pre[1:]
	post = post[:len(post)-1]

	leftLen := 0
	for i, val := range post {
		if pre[0] == val {
			leftLen = i + 1
			break
		}
	}
	leftPre := pre[:leftLen]
	rightPre := pre[leftLen:]
	leftPost := post[:leftLen]
	rightPost := post[leftLen:]

	root.Left = constructFromPrePost(leftPre, leftPost)
	root.Right = constructFromPrePost(rightPre, rightPost)

	return root
}

File: src/#889/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

var constructFromPrePost = function (pre, post) {
  if (pre.length === 0) {
    return null;
  }

  let root = new TreeNode(pre[0]);

  if (pre.length === 1) {
    return root;
  }

  pre.shift();
  post.pop();
  let leftLen = 0;
  post.forEach((v, i) => {
    if (v === pre[0]) {
      leftLen = i + 1;
    }
  });
  let leftPre = pre.slice(0, leftLen);
  let rightPre = pre.slice(leftLen);
  let leftPost = post.slice(0, leftLen);
  let rightPost = post.slice(leftLen);

  root.left = constructFromPrePost(leftPre, leftPost);
  root.right = constructFromPrePost(rightPre, rightPost);

  return root;
};
